pkg/base/logger/runtime_formatter.go

```diff
@@ -0,0 +1,61 @@
+package logger
+
+import (
+	"fmt"
+	"runtime"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+)
+
+// FunctionKey holds the function field
+const FunctionKey = "function"
+
+// FileKey holds the file field
+const FileKey = "file"
+
+const (
+	logrusStackJump          = 4
+	logrusFieldlessStackJump = 6
+)
+
+// RuntimeFormatter decorates log entries with function name and package name (optional) and line number (optional)
+type RuntimeFormatter struct {
+	ChildFormatter logrus.Formatter
+	MaxLevel       logrus.Level
+}
+
+// Format the current log entry by adding the function name and line number of the caller.
+func (f *RuntimeFormatter) Format(entry *logrus.Entry) ([]byte, error) {
+	data := logrus.Fields{}
+	if f.MaxLevel >= entry.Level {
+		function, file, line := f.getCurrentPosition(entry)
+		packageEnd := strings.LastIndex(function, ".")
+		functionName := function[packageEnd+1:]
+
+		data[FunctionKey] = functionName
+		data[FileKey] = file + ":" + line
+	}
+	for k, v := range entry.Data {
+		data[k] = v
+	}
+	entry.Data = data
+
+	return f.ChildFormatter.Format(entry)
+}
+
+func (f *RuntimeFormatter) getCurrentPosition(entry *logrus.Entry) (string, string, string) {
+	skip := logrusStackJump
+	if len(entry.Data) == 0 {
+		skip = logrusFieldlessStackJump
+	}
+start:
+	pc, file, line, _ := runtime.Caller(skip)
+	lineNumber := fmt.Sprintf("%d", line)
+	function := runtime.FuncForPC(pc).Name()
+	if strings.LastIndex(function, "sirupsen/logrus.") != -1 {
+		skip++
+		goto start
```

Comment:
Goto is done to 6 lines higher. So, a bit of side information here:
1. Goto is a very efficient way of writing code - it's just a simple instruction, that is one of easiest from the productivity of view at all.
2. It's a common way of writing code, especially in C language and is used particularly in every well-known system/application.
3. It's a good practice (and yes - forget what teachers says at the univercity :)) to use this mechanism in application - BUT if you use this carefully.

So, it can be used here, because code is very easy to be read, there are no performance overhead at all,

Reply:
1. indeed
2. C != Go
3. indeed

This code can be implemented without `goto`.

Using `runtime` package leads to performance issue.

Reply:
And this approach without goto is more easy to read/understand/more efficient?

We need to use runtime to complete this requirements - it was discussed at friday meeting.